### PR TITLE
24 requiring unique esdl asset id instead of name

### PR DESCRIPTION
**Pull request by FJanssen-TNO**
Updating the code to use esdl asset id's instead of the asset name. Thereby the requirement for unique asset names is no longer needed.

- [x] code change from asset name to asset id
- [x] updating all test cases
- [x] adding test case that name duplicates are now allowed and used correctly as long as the ids are unique.

Only to be merged to code after 0.9 release of NWN.

### CODE REVIEW

File: src/mesido/esdl/profile_parser.py [hunk 1]
```diff
@@ -108,7 +108,7 @@ def read_profiles(
             for component_type, var_name in self.component_type_to_var_name_map.items():
                 for component in energy_system_components.get(component_type, []):
                     profile = self._profiles[ensemble_member].get(component + var_name, None)
-                    asset_power = esdl_assets[esdl_asset_names_to_ids[component]].attributes[
+                    asset_power = esdl_assets[component].attributes[
```
Comment (KobusVanRooyen) [inline src/mesido/esdl/profile_parser.py]: "component" is now an asset id, so the warning print out to the terminal is now difficult to read since the name of the asset is not shown in the message but rather the asset id. See the warning message in read_profiles() -> "f"No profile provided for {component=}" ..." as an example.

The are probably other places that this will occur as well, also error message will be impacted.

I do prefer to see the asset name in these warning/error message.

**Comment by FJanssen-TNO** [comment]
@KobusVanRooyen ready for review again.

File: src/mesido/esdl/esdl_additional_vars_mixin.py [hunk 2]
```diff
@@ -178,7 +178,7 @@ def __limit_list_available_heat_pipe_classes_max_power(
             *self.energy_system_components.get("heat_buffer", []),
             *self.energy_system_components.get("heat_pump", []),
         ]:
-            esdl_asset = self.esdl_assets[self.esdl_asset_name_to_id_map[asset]]
+            esdl_asset = self.esdl_assets[asset]
```
Comment (KobusVanRooyen) [inline src/mesido/esdl/esdl_additional_vars_mixin.py:181]: In this for loop "asset" is now an id, and the error message has not been update to show the asset name:

logger.error(
                            f"{asset} has a setpoint constaint specified with unknown"
                            f"per time unit"
                        )

File: src/mesido/esdl/profile_parser.py [hunk 3]
```diff
@@ -366,7 +362,7 @@ def _load_profiles_from_source(
 
             if isinstance(container, esdl.ProfileConstraint):
                 variable_suffix = self.asset_type_to_variable_name_conversion[type(asset)]
-                var_base_name = asset.name
+                var_base_name = asset.id
```
Comment (KobusVanRooyen) [inline src/mesido/esdl/profile_parser.py:365]: "var_base_name":
- this variable name was used to represent a "name" but now it represents and id. I would suggest changing this variable name to "var_base_id"
- the logger.error has not been updated to show the name as before.

The items above at 2 places. Here and after "elif isinstance(container, esdl.Port)" in the code below

File: src/mesido/asset_sizing_mixin.py [hunk 4]
```diff
@@ -803,7 +803,11 @@ def _make_max_size_var(name, lb, ub, nominal):
             ub = bounds[f"{asset_name}.Heat_source"][1]
 
             # Update bound to account for profile constraint being used instead of 1 value
-            esdl_asset_attributes = self.esdl_assets[asset_name].attributes["constraint"]
+
```
Comment (KobusVanRooyen) [inline src/mesido/asset_sizing_mixin.py:14]: In this file (and probably many others) there are variable names (see the for loops) where the variable name "asset_name" is being used to represent an id of an asset. I think this has to be updated? What do you think. I really think it gets confusing when a variable name represents something that is not expected

File: src/mesido/esdl/esdl_parser.py [hunk 5]
```diff
@@ -131,19 +131,18 @@ def read_esdl(self) -> None:
                 if isinstance(el, esdl.Asset):
                     if hasattr(el, "name") and el.name:
                         el_name = el.name
+                        el_id = el.id
                     else:
-                        el_name = el.id
+                        el_name = el_id = el.id
```
Comment (KobusVanRooyen) [inline src/mesido/esdl/esdl_parser.py:136]: I think this variable name is confusing since it is not a name but an id

File: src/mesido/esdl/profile_parser.py [hunk 6]
```diff
@@ -106,31 +106,59 @@ def read_profiles(
...
+
                     if profile is not None:
                         values = profile
                     else:
                         if "heat_demand" not in component_type:
                             # We don't set a default profile for source targets
                             continue
                         logger.warning(
-                            f"No profile provided for {component=} and "
+                            f"No profile provided for {component_id=} and "
```
Comment (KobusVanRooyen) [inline src/mesido/esdl/profile_parser.py:116]: Can we show the asset name here as before?

File: src/mesido/physics_mixin.py [hunk 7]
```diff
@@ -162,7 +160,7 @@ def __setpoint_constraint(
                     comptype
                     for comptype, compnames in self.energy_system_components.items()
                     for compname in compnames
-                    if compname == component_name
+                    if compname == component_id
```
Comment (KobusVanRooyen) [inline src/mesido/physics_mixin.py]: can we replace "compname" so that "name" is not present in the variable name, to prevent confusing

File: src/mesido/workflows/multicommodity_simulator_workflow.py [hunk 8]
```diff
@@ -588,7 +588,7 @@ def __merit_controls(self, assets_list):
                 if attributes["merit_order"][-1] <= 0.0:
                     raise Exception(
                         "The specified producer usage merit order must be a "
-                        f"positve integer value, producer name:{a.name}, current "
+                        f"positve integer value, producer name:{a.id}, current "
```
Comment (KobusVanRooyen) [inline src/mesido/workflows/multicommodity_simulator_workflow.py:558]: id is still used here instead of name

File: tests/test_roll_out.py [hunk 9]
```diff
@@ -50,11 +50,13 @@ class RollOutTimeStep(RollOutProblem):
             def read(self):
                 super().read()
                 # m = [3, 5, 5]
-                m = [1, 2, 2]
-                for i in range(1, 4):
-                    demand_timeseries = self.get_timeseries(f"HeatingDemand_{i}.target_heat_demand")
-                    demand_timeseries.values[:] = demand_timeseries.values[:] * m[i - 1]
-                    self.set_timeseries(f"HeatingDemand_{i}.target_heat_demand", demand_timeseries)
+                m = [1.5, 1.5, 1.5]
```
Comment (KobusVanRooyen) [inline tests/test_roll_out.py:53]: why did the numbers now change?

File: tests/test_end_scenario_sizing.py [hunk 10]
```diff
@@ -515,41 +516,41 @@ def test_end_scenario_sizing_pipe_catalog(self):
...
-        original_problem_pipe_class_cost_map = {
-            org_pipe.name: org_pipe.investment_costs
-            for org_pipe in original_problem_pipe_classes
-            if not org_pipe.name == "None"
-        }
-        np.testing.assert_equal(
-            solution_pipe_class_cost_map.items() != original_problem_pipe_class_cost_map.items(),
-            True,
-        )
+        #
```
Comment (KobusVanRooyen) [inline tests/test_end_scenario_sizing.py:444]: Why has this been commented out?

File: tests/test_esdl_parsing.py [hunk 11]
```diff
@@ -48,6 +48,71 @@ def test_from_string_and_from_file_are_equal(self):
...
+        ), f"Pipe {pipe_1_id} not found in energy_system_components heat_pipe list"
+        assert (
+            pipe_2_id in heat_pipes
+        ), f"Pipe {pipe_2_id} not found in energy_system_components heat_pipe list"
+
+        # Verify that both pipes have the same name in the ESDL
+        pipe_1_asset = problem.esdl_assets[pipe_1_id]
+        pipe_2_asset = problem.esdl_assets[pipe_2_id]
+
+        assert (
```
Comment (KobusVanRooyen) [inline tests/test_esdl_parsing.py:95]: Why are asserts used instead of np.testing....

**Comment by KobusVanRooyen** [comment]
@FJanssen-TNO review completed

File: tests/test_roll_out.py (hunk 9 above)
Comment (FJanssen-TNO) [reply tests/test_roll_out.py:53]: This was a wrongly picked numbering as the order of assets can be different in the list. So far we had been lucky. Everything that was required in the testing still works with these numbers.

File: tests/test_esdl_parsing.py (hunk 11 above)
Comment (FJanssen-TNO) [reply tests/test_esdl_parsing.py:95]: They have the same purpose, it just felt strange to compare strings with numpy.

File: src/mesido/asset_sizing_mixin.py (hunk 4 above)
Comment (FJanssen-TNO) [reply src/mesido/asset_sizing_mixin.py:14]: I was not planning to update it now. 
Also sometimes also only the word "asset" is used. Since many of them are loops over a list of assets and asset names are no longer to be used int he code, all asset_names except in parsing are now asset_ids. 
If we were to do this consistent everywhere, either we can change it on the go, in pieces. Or we change it all at onces preferably in a separate PR.

File: src/mesido/workflows/multicommodity_simulator_workflow.py (hunk 8 above)
Comment (FJanssen-TNO) [reply src/mesido/workflows/multicommodity_simulator_workflow.py:558]: Sorry I don't understand what you mean.

File: src/mesido/esdl/profile_parser.py (hunk 6 above)
Comment (FJanssen-TNO) [reply src/mesido/esdl/profile_parser.py:116]: what do you mean with here? The line you selected is about the asset_power.

File: src/mesido/esdl/profile_parser.py (hunk 3 above)
Comment (FJanssen-TNO) [reply src/mesido/esdl/profile_parser.py:365]: - I do not agree with you in the first point. "var_base_name" is an abbreviation of "variable base name" whether that is an asset id or asset name or something else, is independent. 
- The second point I agree, will be updated.

File: src/mesido/workflows/multicommodity_simulator_workflow.py (hunk 8 above)
Comment (KobusVanRooyen) [reply src/mesido/workflows/multicommodity_simulator_workflow.py:558]: Sorry ,this message was placed in the wrong place. Please see the Exception --> "positve integer value, producer name:{a.id},"

File: src/mesido/esdl/esdl_parser.py (hunk 5 above)
Comment (FJanssen-TNO) [reply src/mesido/esdl/esdl_parser.py:136]: I think I agree but from here on the id is used as the identifier and thus as the variable name.

File: src/mesido/esdl/profile_parser.py (hunk 6 above)
Comment (KobusVanRooyen) [reply src/mesido/esdl/profile_parser.py:116]: f"No profile provided for {component_id=} and " ---> previously the name was used in the warning message, and now it is an id

File: src/mesido/esdl/profile_parser.py (hunk 3 above)
Comment (KobusVanRooyen) [reply src/mesido/esdl/profile_parser.py:365]: I agree

File: tests/test_end_scenario_sizing.py (hunk 10 above)
Comment (FJanssen-TNO) [reply tests/test_end_scenario_sizing.py:444]: Oeps, when I was updating the code, I temporarily commented it out. Forgot to get it back in. But have restored it now :)

File: src/mesido/esdl/esdl_additional_vars_mixin.py (hunk 2 above)
Comment (FJanssen-TNO) [reply src/mesido/esdl/esdl_additional_vars_mixin.py:181]: updated.

File: src/mesido/esdl/profile_parser.py (hunk 6 above)
Comment (FJanssen-TNO) [reply src/mesido/esdl/profile_parser.py:116]: updated.

File: src/mesido/workflows/multicommodity_simulator_workflow.py (hunk 8 above)
Comment (FJanssen-TNO) [reply src/mesido/workflows/multicommodity_simulator_workflow.py:558]: updated.

File: src/mesido/physics_mixin.py (hunk 7 above)
Comment (FJanssen-TNO) [reply src/mesido/physics_mixin.py]: updated

**Comment by KobusVanRooyen** [comment]
@FJanssen-TNO approved
